sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -231,14 +231,15 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
     val keysAndIndexes = currentOrderOfKeys.zipWithIndex
 
     expectedOrderOfKeys.foreach(expression => {
-      val index = keysAndIndexes.find { case (e, idx) =>
+      keysAndIndexes.find { case (e, idx) =>
         // As we may have the same key used many times, we need to filter out its occurrence we
         // have already used.
         e.semanticEquals(expression) && !pickedIndexes.contains(idx)
-      }.map(_._2).get
-      pickedIndexes += index
-      leftKeysBuffer.append(leftKeys(index))
-      rightKeysBuffer.append(rightKeys(index))
+      }.map(_._2).map(index => {
```

Comment:
This has to be `foreach`, not `map`, and can be `.foreach { index =>`
However I wonder if there's some other bug if the code really doesn't expect to not find a value?

Reply:
I'll check it.

Reply:
If `Set(currentOrderOfKeys) == Set(expectedOrderOfKeys)`, I think we cannot hit the exception...

Reply:
> If `Set(currentOrderOfKeys) == Set(expectedOrderOfKeys)`, I think we cannot hit the exception...

yes,  you are right.

Reply:
We can add a judgement in reorderJoinKeys.